File: hw/aimacana/u3/hw22MazeSolidPrinciplesCode/MazeJavaScript/view/MazePanel.js
export default class MazePanel {
    constructor(maze, path, cellSize, canvasId) {
        this.maze = maze;
        this.path = path;
        this.cellSize = cellSize;
        this.canvas = document.getElementById(canvasId);
        if (!this.canvas) {
            console.error(`Canvas with id ${canvasId} not found`);
            return;
        }
        
        const w = maze.width * cellSize;
        const h = maze.height * cellSize;
        this.canvas.width = w + cellSize;
        this.canvas.height = h + cellSize;
        this.ctx = this.canvas.getContext('2d');
    }

    paintComponent() {
        if (!this.ctx) return;
        
        const width = this.maze.width;
        const height = this.maze.height;
        const margin = this.cellSize / 2;

        const wallColor = "#4a5568";
        const pathColor = "#ef4444";
        const entranceColor = "#22c55e";
        const exitColor = "#3b82f6";
        const cellBgColor = "#1e293b";
        const bgColor = "#121218";

        // Fill canvas background
        this.ctx.fillStyle = bgColor;
        this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);

        // Fill maze background
        this.ctx.fillStyle = cellBgColor;
        this.ctx.fillRect(margin, margin, width * this.cellSize, height * this.cellSize);

        const entrance = this.maze.entrance;
        if (entrance) {
            this.ctx.fillStyle = entranceColor;
            this.ctx.fillRect(margin + entrance.x * this.cellSize, margin + entrance.y * this.cellSize, this.cellSize, this.cellSize);
        }
        
        const exit = this.maze.exit;
        if (exit) {
            this.ctx.fillStyle = exitColor;
            this.ctx.fillRect(margin + exit.x * this.cellSize, margin + exit.y * this.cellSize, this.cellSize, this.cellSize);
        }

        if (this.path && this.path.length > 0) {
            this.ctx.strokeStyle = pathColor;
            this.ctx.lineWidth = Math.max(2, this.cellSize / 4);
            this.ctx.lineCap = 'round';
            this.ctx.lineJoin = 'round';

            this.ctx.beginPath();
            for (let i = 0; i < this.path.length - 1; i++) {
                const c1 = this.path[i];
                const c2 = this.path[i + 1];
                const x1 = margin + c1.x * this.cellSize + this.cellSize / 2;
                const y1 = margin + c1.y * this.cellSize + this.cellSize / 2;
                const x2 = margin + c2.x * this.cellSize + this.cellSize / 2;
                const y2 = margin + c2.y * this.cellSize + this.cellSize / 2;
                
                this.ctx.moveTo(x1, y1);
                this.ctx.lineTo(x2, y2);
            }
            this.ctx.stroke();
        }

        this.ctx.strokeStyle = wallColor;
        this.ctx.lineWidth = Math.max(1, this.cellSize / 10);

        this.ctx.beginPath();
        for (let y = 0; y < height; y++) {
            for (let x = 0; x < width; x++) {
                const cell = this.maze.getCell(x, y);
                if (!cell) continue;

                const cx = margin + x * this.cellSize;
                const cy = margin + y * this.cellSize;

                if (cell.wallNorth) {
                    this.ctx.moveTo(cx, cy);
                    this.ctx.lineTo(cx + this.cellSize, cy);
                }
                if (cell.wallSouth) {
                    this.ctx.moveTo(cx, cy + this.cellSize);
                    this.ctx.lineTo(cx + this.cellSize, cy + this.cellSize);
                }
                if (cell.wallEast) {
                    this.ctx.moveTo(cx + this.cellSize, cy);
                    this.ctx.lineTo(cx + this.cellSize, cy + this.cellSize);
                }
                if (cell.wallWest) {
                    this.ctx.moveTo(cx, cy);
                    this.ctx.lineTo(cx, cy + this.cellSize);
                }
            }
        }
        this.ctx.stroke();
    }
}

File: hw/aimacana/u3/hw22MazeSolidPrinciplesCode/MazePython/view/maze_panel.py
import tkinter as tk
from model.maze import Maze

class MazePanel(tk.Canvas):
    def __init__(self, master, maze: Maze, path: list, cell_size: int):
        self.maze = maze
        self.path = path
        self.cell_size = cell_size
        
        w = maze.width * cell_size
        h = maze.height * cell_size
        
        super().__init__(master, width=w + cell_size, height=h + cell_size, bg="#121218", highlightthickness=0)
        self.bind("<Configure>", self.on_paint)
        
    def on_paint(self, event=None):
        self.delete("all")
        width = self.maze.width
        height = self.maze.height
        margin = self.cell_size // 2
        
        # Colors
        wall_color = "#4a5568"
        path_color = "#ef4444"
        entrance_color = "#22c55e"
        exit_color = "#3b82f6"
        cell_bg_color = "#1e293b"
        
        # Background
        self.create_rectangle(margin, margin, margin + width * self.cell_size, margin + height * self.cell_size, fill=cell_bg_color, outline="")
        
        # Entrance / Exit
        entrance = self.maze.entrance
        if entrance:
            ex = margin + entrance.x * self.cell_size
            ey = margin + entrance.y * self.cell_size
            self.create_rectangle(ex, ey, ex + self.cell_size, ey + self.cell_size, fill=entrance_color, outline="")
            
        exit_cell = self.maze.exit
        if exit_cell:
            ex = margin + exit_cell.x * self.cell_size
            ey = margin + exit_cell.y * self.cell_size
            self.create_rectangle(ex, ey, ex + self.cell_size, ey + self.cell_size, fill=exit_color, outline="")
            
        # Path
        if self.path:
            line_width = max(2, self.cell_size // 4)
            for i in range(len(self.path) - 1):
                c1 = self.path[i]
                c2 = self.path[i + 1]
                x1 = margin + c1.x * self.cell_size + self.cell_size // 2
                y1 = margin + c1.y * self.cell_size + self.cell_size // 2
                x2 = margin + c2.x * self.cell_size + self.cell_size // 2
                y2 = margin + c2.y * self.cell_size + self.cell_size // 2
                self.create_line(x1, y1, x2, y2, fill=path_color, width=line_width, capstyle=tk.ROUND, joinstyle=tk.ROUND)
                
        # Walls
        wall_width = max(1, self.cell_size // 10)
        for y in range(height):
            for x in range(width):
                cell = self.maze.get_cell(x, y)
                if not cell:
                    continue
                
                cx = margin + x * self.cell_size
                cy = margin + y * self.cell_size
                
                if cell.is_wall_north:
                    self.create_line(cx, cy, cx + self.cell_size, cy, fill=wall_color, width=wall_width)
                if cell.is_wall_south:
                    self.create_line(cx, cy + self.cell_size, cx + self.cell_size, cy + self.cell_size, fill=wall_color, width=wall_width)
                if cell.is_wall_east:
                    self.create_line(cx + self.cell_size, cy, cx + self.cell_size, cy + self.cell_size, fill=wall_color, width=wall_width)
                if cell.is_wall_west:
                    self.create_line(cx, cy, cx, cy + self.cell_size, fill=wall_color, width=wall_width)
